### PR TITLE
Release Pipeline

### PR DESCRIPTION
**Pull request by jamiebrynes7**
Add a release pipeline for our entire release. The general flow is:

1. Kick off release pipeline
2. Fill in version info.
3. Pipeline creates RC branch.
4. Pipeline kicks off release QA build (this is injected by a shell script).
5. Once you do release QA, unblock the pipeline.
6. Pipeline merges the RC PR and drafts the release.

Repeat 3-6 for the remaining repositories.

This process is entirely containerized and uses @gdk-for-unity-bot's credentials for all operations. The SSH key and Github token are injected via `imp-ci` and are cleaned up at the end of the build.

Some build steps need to add metadata back from the Docker container to the build pipeline (for example: PR URLs, branch names). This is done by dumping the key value pairs into a well known file which is in a folder mounted into the Docker container. We can then read this file and call `buildkite-agent meta-data set` without injecting all of that into the build.

#### Remaining work

- [x] Bake Linux agent with Docker.
- [x] Test flow (we can revert the PRs that it merges 😄)
- [x] Add golden build queue

### CODE REVIEW

File: .buildkite/release.definition.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+agent_queue_id: trigger-pipelines
+description: Add a description...
```
Comment (zeroZshadow) [inline .buildkite/release.definition.yaml]: Add a description

File: .buildkite/release.definition.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+github:
+  branch_configuration: []
+  default_branch: master
+  pull_request_branch_filter_configuration: []
+teams:
+- name: Everyone
```
Comment (zeroZshadow) [inline .buildkite/release.definition.yaml:8]: I know little of this, but should this be "Everyone"?

File: ci/common.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+function writeBuildkiteMetadata() {
+    BK_METADATA_FILE="${1}"
+
+    # If we want to upload any metadata elements. Do so here.
+    if [[ -f "${BK_METADATA_FILE}" ]]; then
+        while IFS= read -r line
+        do
+            KEY=$(echo ${line} | cut -d "," -f 1)
```
Comment (zeroZshadow) [inline ci/common.sh]: line needs quotes

File: ci/prepare-release.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+
+### This script should only be run on Improbable's internal build machines.
+### If you don't work at Improbable, this may be interesting as a guide to what software versions we use for our
+### automation, but not much more than that.
+
+set -e -u -x -o pipefail
+
+if [ -z "$BUILDKITE" ]; then
```
Comment (zeroZshadow) [inline ci/prepare-release.sh]: double [[ ?

File: ci/prepare-release.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  echo "This script is only to be run on Improbable CI."
+  echo 0
```
Comment (zeroZshadow) [inline ci/prepare-release.sh]: echo not exit?

File: ci/prepare-release.sh [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	PIN_HASH="$(buildkite-agent meta-data get gdk-for-unity-hash)"
+	PIN_ARG="--update-pinned-gdk=${PIN_HASH}"
+else
+	PIN_ARG=""
+fi
+
+docker run \
+	-v "${SECRETS_DIR}":/var/ssh \
+	-v "${SECRETS_DIR}":/var/github \
+	-v $(pwd)/logs:/var/logs \
```
Comment (zeroZshadow) [inline ci/prepare-release.sh]: quotes

File: ci/prepare-release.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	local:gdk-release-tool \
+		prep "${RELEASE_VERSION}" \
+			--git-repository-name="${REPO}" \
+			--github-key-file="/var/github/github_token" \
+			--buildkite-metadata-path="/var/logs/bk-metadata" ${PIN_ARG}
```
Comment (zeroZshadow) [inline ci/prepare-release.sh]: quotes?

File: ci/release.sh [hunk 8]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+
+### This script should only be run on Improbable's internal build machines.
+### If you don't work at Improbable, this may be interesting as a guide to what software versions we use for our
+### automation, but not much more than that.
+
+set -e -u -x -o pipefail
+
+if [ -z "$BUILDKITE" ]; then
```
Comment (zeroZshadow) [inline ci/release.sh]: You know what to do here

File: ci/release.sh [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+PR_URL="$(buildkite-agent meta-data get ${REPO}-pr-url)"
+
+setupReleaseTool
+
+mkdir -p ./logs
+
+docker run \
+	-v "${SECRETS_DIR}":/var/ssh \
+	-v "${SECRETS_DIR}":/var/github \
+	-v $(pwd)/logs:/var/logs \
```
Comment (zeroZshadow) [inline ci/release.sh]: Quotes

File: tools/ReleaseTool/BuildkiteMetadataSink.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        public interface IBuildkiteOptions
+        {
+            [Option("buildkite-metadata-path", HelpText = "The location of a file which we can write metadata key-pairs to.")]
+            string MetadataFilePath { get; set; }
+        }
+
+        public static bool CanWrite(IBuildkiteOptions options)
+        {
+            return options.MetadataFilePath != null;
```
Comment (zeroZshadow) [inline tools/ReleaseTool/BuildkiteMetadataSink.cs]: string.NullOrEmpty

File: tools/ReleaseTool/BuildkiteMetadataSink.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        private StreamWriter outputFile;
+
+        public BuildkiteMetadataSink(IBuildkiteOptions options)
+        {
+            if (options.MetadataFilePath == null)
```
Comment (zeroZshadow) [inline tools/ReleaseTool/BuildkiteMetadataSink.cs]: see above

File: tools/ReleaseTool/ReleaseCommand.cs [hunk 12]
```diff
@@ -70,7 +72,7 @@ public int Run()
                     throw new InvalidOperationException(
                         $"Was unable to merge pull request at: {options.PullRequestUrl}. Received error: {mergeResult.Message}");
                 }
-
+
```
Comment (zeroZshadow) [inline tools/ReleaseTool/ReleaseCommand.cs]: whitespace

**Review by zeroZshadow — CHANGES_REQUESTED** [review]

File: ci/prepare-release.sh (hunk 7 above)
Comment (jamiebrynes7) [reply ci/prepare-release.sh]: Will need to test this, it may interfere with the arg parsing

**Review by zeroZshadow — APPROVED** [review]

File: ci/prepare-release.sh (hunk 7 above)
Comment (jamiebrynes7) [reply ci/prepare-release.sh]: Actually its previously quoted so its fine 😄

File: .buildkite/hooks/pre-command [hunk 13]
```diff
@@ -2,6 +2,6 @@
 set -euo pipefail
 [[ -n "${DEBUG:-}" ]] && set -x
 
-if [[ "${BUILDKITE_AGENT_META_DATA_CAPABLE_OF_BUILDING}" == "gdk-for-unity" ]]; then
+if [[ "${BUILDKITE_AGENT_META_DATA_CAPABLE_OF_BUILDING}" == "gdk-for-unity" ]] && [[ "${BUILDKITE_AGENT_META_DATA_OS}" == "windows" ]]; then
```
Comment (jared-improbable) [inline .buildkite/hooks/pre-command:5]: No lunix for u?

File: ci/docker/release-tool.Dockerfile [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+# Setup GIT
+RUN apt-get update && \
+    apt-get upgrade -y && \
+    apt-get install -y git && \
+    git config --global user.name "GDK for Unity Bot" && \
+    git config --global user.email "gdk-for-unity-bot@improbable.io" && \
+    git config --global core.sshCommand "ssh -i /var/ssh/id_rsa" && \
+    mkdir -p /root/.ssh && \
+    touch /root/.ssh/known_hosts && \
+    ssh-keyscan github.com >> /root/.ssh/known_hosts
```
Comment (jared-improbable) [inline ci/docker/release-tool.Dockerfile:22]: Nice.

File: tools/ReleaseTool/BuildkiteMetadataSink.cs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        public static bool CanWrite(IBuildkiteOptions options)
+        {
+            return string.IsNullOrEmpty(options.MetadataFilePath);
+        }
+
+        private StreamWriter outputFile;
+
+        public BuildkiteMetadataSink(IBuildkiteOptions options)
+        {
+            if (string.IsNullOrEmpty(options.MetadataFilePath))
```
Comment (jared-improbable) [inline tools/ReleaseTool/BuildkiteMetadataSink.cs]: `if (!CanWrite())`?

File: tools/ReleaseTool/PrepCommand.cs [hunk 16]
```diff
@@ -35,17 +35,16 @@ internal class PrepCommand
...
             public string Version { get; set; }
 
             [Option('g', "update-pinned-gdk", HelpText =
                 "The git hash of the version of the gdk to upgrade to. (Only if this is a project).")]
             public string PinnedGdkVersion { get; set; }
-
-            // TODO: Once we have a robots account - set this to default robot account.
-            [Option("github-user", HelpText = "The user account that is using this.", Required = true)]
+            
+            [Option("github-user", HelpText = "The user account that is using this.", Default = "gdk-for-unity-bot")]
```
Comment (jared-improbable) [inline tools/ReleaseTool/PrepCommand.cs]: Does this need to be configurable?

File: .buildkite/hooks/pre-command (hunk 13 above)
Comment (jamiebrynes7) [reply .buildkite/hooks/pre-command:5]: No `spatial` on Linux 😉

File: tools/ReleaseTool/PrepCommand.cs (hunk 16 above)
Comment (jamiebrynes7) [reply tools/ReleaseTool/PrepCommand.cs]: Actually not anymore, will fix

File: ci/prepare-release.sh [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+if [[ "${REPO}" != "gdk-for-unity" ]]; then
+	PIN_HASH="$(buildkite-agent meta-data get gdk-for-unity-hash)"
+	PIN_ARG="--update-pinned-gdk=${PIN_HASH}"
+else
+	PIN_ARG=""
+fi
+
+docker run \
+	-v "${SECRETS_DIR}":/var/ssh \
```
Comment (paulbalaji) [inline ci/prepare-release.sh]: using tabs here - should be spaces

File: ci/release.sh [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+REPO="${1}"
+RELEASE_VERSION="$(buildkite-agent meta-data get release-version)"
+PR_URL="$(buildkite-agent meta-data get ${REPO}-pr-url)"
+
+setupReleaseTool
+
+mkdir -p ./logs
+
+docker run \
+	-v "${SECRETS_DIR}":/var/ssh \
```
Comment (paulbalaji) [inline ci/release.sh]: tabs again

**Review by paulbalaji — APPROVED** [review]
LGTM % tabs

**Pull request merged**